packages/@react-stately/combobox/src/useComboBoxState.ts

```diff
@@ -147,15 +177,19 @@ export function useComboBoxState<T extends object>(props: ComboBoxStateProps<T>)
 
   let commitCustomValue = () => {
     let shouldClose = false;
-    if (!allowsCustomValue) {
-      resetInputValue();
-      shouldClose = inputValue === lastValue.current;
-    } else {
-      lastSelectedKey.current = null;
-      setSelectedKey(null);
-      shouldClose = selectedKey === null;
+    lastSelectedKey.current = null;
+    setSelectedKey(null);
+
+    // If previous key was already null, need to manually call onSelectionChange since it won't be triggered by a setSelectedKey call
+    // This allows the application to control whether or not to close the menu on custom value commit
+    if (selectedKey === null && props.onSelectionChange) {
+      props.onSelectionChange(null);
     }
 
+    // Should close menu ourselves if component open state or selected key is uncontrolled and therefore won't be closed by a user defined event handler
+    shouldClose = props.isOpen == null || props.selectedKey === undefined;
```

Comment:
Regarding the question for why `props.isOpen` is required here, this is so we close the menu when submitting a custom value when isOpen is uncontrolled. `commitCustomValue` sets `lastSelectedKey.current` to `null` and calls `setSelectedKey(null)`, making it so that the useEffect's block that handles closing the menu on selected key change doesn't fire since `lastSelectedKey = selectedKey`.